### ISSUE TITLE
Template part has been deleted or is unavailable

### ISSUE THREAD

**Issue by janboddez** [WordPress/gutenberg#44243]
### Description

I've been working on a Twenty Twenty-Two child theme and it's been great.

I then tried to do the same (create a child theme) for Anders Norén's Björk, but the site's broken and the Editor shows a bunch of 'Template part has been deleted or is unavailable' errors.

`debug.log` shows errors like these: 
```
[18-Sep-2022 15:27:33 UTC] REST request: /wp/v2/template-parts/bjork//sidebar: {"context":"edit","id":"bjork\/\/sidebar","_locale":"user"}(Header Present)
[18-Sep-2022 15:27:33 UTC] REST result: /wp/v2/template-parts/(?P<id>([^\/:<>\*\?"\|]+(?:\/[^\/:<>\*\?"\|]+)?)[\/\w-]+): {"code":"rest_template_not_found","message":"Er bestaan geen templates met dat id.","data":{"status":404}}(404) - [](User ID: 1)
[18-Sep-2022 15:27:33 UTC] REST request: /wp/v2/template-parts/bjork//footer: {"context":"edit","id":"bjork\/\/footer","_locale":"user"}(Header Present)
[18-Sep-2022 15:27:33 UTC] REST result: /wp/v2/template-parts/(?P<id>([^\/:<>\*\?"\|]+(?:\/[^\/:<>\*\?"\|]+)?)[\/\w-]+): {"code":"rest_template_not_found","message":"Er bestaan geen templates met dat id.","data":{"status":404}}(404) - [](User ID: 1)
```
Tried installing the Gutenberg plugin, but alas. (The TT2 child theme works great even without it.)

Found https://github.com/WordPress/create-block-theme/issues/27 and https://core.trac.wordpress.org/ticket/54507 but they both seem solved, so I think something else might be going on.

### Step-by-step reproduction instructions

Install Björk. Create (nearly) blank child theme. Like, just a `style.css` with the necessary information in it.

Activate it. Visit site, or Site Editor. You'll see missing block (part) templates.

Expect everything to just work as if the parent theme was active. As has been my experience with (a child theme of) TT2, luckily.

### Environment info

- WordPress 6.0.2
- Gutenberg: 14.1.0
- Theme: Björk 0.1.6

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by annezazu** [comment]
@MaggieCabrera any insight here based on your prior comments on a related issue? 

Also, @janboddez thanks for opening! Can you confirm whether this happens with other block themes other than the Bjork one?

**Comment by janboddez** [comment]
Tried another theme by the same author (Poe) and this time it seems to be only the "loop" block that's "unavailable."

Also tried with a child theme of Blockbase and that seems to work fine, for now (just had a quick glance at a few templates, nothing seemed off).

Did all of this on a different site than the first one, just in case.

So it *might* be a theme issue, but I don't see anything odd about how template parts are defined in `theme.json`. Also, the files are definitely there in the parent's `parts` folder (and work when the parent theme itself is active).

I also noticed that if I go and edit the *Template Part itself*, it's all there, as it should (even with the child theme active).

**Comment by annezazu** [comment]
Hmm by chance, have you tried using this plugin to create a child theme instead? https://wordpress.org/plugins/create-block-theme/ Perhaps @jffng  has some insight here :)

**Comment by janboddez** [comment]
Just did, same result.

**Comment by Drivingralle** [comment]
> Just did, same result.

Can confirm @janboddez report. Used the plugin to create child theme and get same error with Björk.

**Comment by janboddez** [comment]
TL;DR
The issue is with the theme(s), but maybe this could still be documented better?

Some details:

In the original theme's `bjork/templates/front-page.html`, there's references to three template parts:
```
<!-- wp:template-part {"slug":"header","tagName":"header","className":"site-header","layout":{"inherit":true}} /-->
```
The one above works OK.

These two, however, error out:
```
<!-- wp:template-part {"slug":"sidebar","theme":"bjork","tagName":"aside","className":"site-sidebar"} /-->
<!-- wp:template-part {"slug":"footer","theme":"bjork","tagName":"footer","className":"site-footer"} /-->
```
See how they explicitly mention `"theme":"bjork"`? Well, remove that bit, and everything works.

Found out by adding a couple debug statements to the template REST API endpoint. The first template part above (so without explicit mention of the parent theme) ends up getting an ID of `bjorkchildtheme//front-page`, which is found fine (`bjorkchildtheme` is my child theme's slug). The others result in IDs `bjork//sidebar` and `bjork//footer` which lead to an error.

**Comment by janboddez** [comment]
Left a message on the WP.org support forum, too: https://wordpress.org/support/topic/child-theme-and-template-parts-issue/

Not closing the issue, yet, as it could still be caused by theme creation tooling or whatnot.

**Comment by andersnoren** [comment]
I fixed this in Björk (and in my other block themes) by removing `"theme":"bjork"` from the `<!-- wp:template-part --/>` options. 

I couldn't find any documentation for the `"theme":"{theme-slug}"` option for `<!-- wp:template-part --/>`, so I'm not sure what its intended behavior is.

**Comment by janboddez** [comment]
Cool!

I do agree that either WP should be more robust for this, or the documentation improved (or both).

**Comment by annezazu** [comment]
cc @DaisyOlsen and @justintadlock on the documentation front!

**Comment by mandrasch** [comment]
Experienced this  as well with https://wordpress.org/themes/raft/ by themeislel, `Template part has been deleted or is unavailable: header` when I use a child theme.

Suggestion of @andersnoren to change 

`<!-- wp:template-part {"slug":"header","tagName":"header","theme":"raft"} /-->`

 to 
 
 `<!-- wp:template-part {"slug":"header","tagName":"header"} /-->`
 
  in `wp-content/themes/raft/templates/front-page.html` solved this for the index page.

**Comment by Thelmachido** [comment]
> Hmm by chance, have you tried using this plugin to create a child theme instead? https://wordpress.org/plugins/create-block-theme/ Perhaps @jffng has some insight here :)

The issue was recently raised in forums for this plugin, and the issue was [reported here](https://github.com/WordPress/create-block-theme/issues/242) @matiasbenedetto fixed this 🎊 

 @janboddez can you please check your site if child themes created with the Create Block Plugin are still returning the error? 

The issue has been marked as stale and will likely close soon.

**Comment by janboddez** [comment]
Then there are (were?) two issues:
- Themes exported with that plugin (used to?) sometimes contain spaces, which breaks them (when used on other sites)
- **Child** themes of a theme with _any_ `theme` property/slug in one or several of their templates break (or at least did when I filed this issue, I've since stopped trying to create block child themes anymore); fixed by removing all mentions of a `theme` property

**Comment by matiasbenedetto** [comment]
I made some testing about the template-part loading and this is what I found. **I could reproduce the issue (_Template part has been deleted or is unavailable_) ONLY when the template part is linked from a pattern.** The problem can be reproduced in both stand-alone/parent themes, and child themes.

## Option working as expected:

**To avoid this problem you need to add the `theme` key in the template-parts linked from patterns to make it work.** The theme key should contain the slug of the theme in case of stand-alone themes or the child theme slug in case of the child themes. Any other option (not including the `theme` key, or adding the parent slug when you are on a child theme) won't work and will throw the error: _Template part has been deleted or is unavailable_.

```
<!-- wp:template-part {"slug":"header","theme":"themeslug"} /-->
```

I'm not sure if this was designed to work like this but by doing that I was able to avoid the problem.


## About this issue Create Block Theme Plugin
Exporting stand-alone/parent themes works as expected.
Exporting child themes can fail in some cases because we are using the parent theme slug in the template parts linked from patterns. We need to fix that. I created an issue to follow-up: https://github.com/WordPress/create-block-theme/issues/275


## Extra Info:
If the template part is linked from a template and not from a pattern it works seamlessly no matter if you add the theme key in the link or not.

```
<!-- wp:template-part {"slug":"header","theme":"themeslug"} /-->
<!-- wp:template-part {"slug":"header","theme":"childthemeslug"} /-->
<!-- wp:template-part {"slug":"header"} /-->
```
:point_up: These 3 options are working as expected **when you link a template-part from a template (not from a pattern).**





### Steps to reproduce the issue:
If you want to reproduce the error loading template-parts linked from a pattern you can do something like this:

1. Create an empty template `front-page.html`
2. Create an empty pattern `blog.php`
3. Create a template part `header.html` and some blocks to it.
```
<!-- wp:paragraph -->
<p class="has-text-align-left">Hi I'm a paragraph</p>
<!-- /wp:paragraph -->
```

4. Link the `blog.php` from the `front-page.html` template:
```
<!-- wp:pattern {"slug":"mytheme/blog"} /-->
```
5. Link the `header.html` template part from the `blog.php`pattern.
```
<?php
/**
 * Title: Blog post list
 * Slug: mytheme/blog
 */
?>

<!-- wp:template-part {"slug":"header"} /-->
```

I tested using:
- WordPress 6.1
- Gutenberg 15.3.1

**Comment by Retrofitterdk** [comment]
As far as I can tell this issue is not specifically related to neither patterns nor the Create Block Theme plugin.

From my experience whenever you use the site editor to make changes to a parent theme it will add theme key/themeslug to the wp:template-part html-comment in parent theme templates, and this will in my experience break the child theme.

To recreate the issue in it’s simplest form I have tried with these steps.

1.  Use Create Block Theme plugin to create a blank parent theme.
2. Activate parent theme
3. Observe that the parent theme has created a index template with a link to a template-part without themeslug
`<!-- wp:template-part {"slug":"header","tagName":"header"} /-->`
4. Use Create Block Theme plugin to ’Create child of Parent’.
5. Upload and activate child theme 
6. Observe that Create Block Theme plugin hasn’t created any template or template-parts in the child theme – only style.css and theme.json
7. Observe that the child theme correctly loads templates and template-parts from the parent theme.
8. Reactivate the parent theme
9. Use the site editor to make a change in a template or template-part
10. In the site editor open the index.html template in code editor-’mode’ and observe that the site editor has changed the template-part link to
`<!-- wp:template-part {"slug":"header","theme":"parent","tagName":"header"} /-->`
11. Use Create Block Theme plugin to ’Overwrite Parent’ 
12. In a code editor observe that the index.html template in the parent theme is now changed, so it links to the template-part with
`<!-- wp:template-part {"slug":"header","theme":"parent","tagName":"header"} /-->`
13. Reactivate the child theme 
14. Observe that the child no longer loads the header template part from the parent theme
15. Open the index.html template in the site editor and observe the error message ’Template part has been deleted or is unavailable’.
16. In a code editor manually remove the theme key from the index.html template so the template-part link is restored to
`<!-- wp:template-part {"slug":"header","tagName":"header"} /-->`
17. Observe that the child now correctly loads the header template part from the parent theme

I tested using 

- WordPress 6.2
- Create Block Theme 1.8.1

No other plugins installed

**Comment by janboddez** [comment]
Not sure what happened here; I purposely left out the `theme` attribute of all my pattern (or template) calls and it worked wonderfully for months. Yet when I checked the Site Editor just now, this issue’s back.

Edit: It may very well be due to the fact that my site seems to have upgraded to 6.4 beta 😅

**Comment by t-hamano** [comment]
I followed [this comment](https://github.com/WordPress/gutenberg/issues/44243#issuecomment-1512596814) and tested to see if the problem still occurs. This problem doesn't seem to reproduce in my environment, but how about your environment?

I tested using

- WordPress 6.4.1 (without Gutenberg Plugin)
- Create Block Theme 1.13.4

> 1.  Use Create Block Theme plugin to create a blank parent theme.
> 2. Activate parent theme
> 3. Observe that the parent theme has created a index template with a link to a template-part without themeslug
> `<!-- wp:template-part {"slug":"header","tagName":"header"} /-->`

The template part now had the parent theme's slug:

`<!-- wp:template-part {"slug":"header", "theme": "parent", "tagName":"header"} /-->`

> 4. Use Create Block Theme plugin to ’Create child of Parent’.
> 5. Upload and activate child theme 
> 6. Observe that Create Block Theme plugin hasn’t created any template or template-parts in the child theme – only style.css and theme.json
> 7. Observe that the child theme correctly loads templates and template-parts from the parent theme.

Here, the theme slug was automatically changed to the child theme:

`<!-- wp:template-part {"slug":"header", "theme": "child" "tagName":"header"} /-->`

> 8. Reactivate the parent theme
> 9. Use the site editor to make a change in a template or template-part
> 10. In the site editor open the index.html template in code editor-’mode’ and observe that the site editor has changed the template-part link to
> `<!-- wp:template-part {"slug":"header","theme":"parent","tagName":"header"} /-->`

Before updating the template, the theme slug had changed to that of the parent:

`<!-- wp:template-part {"slug":"header", "theme": "parent", "tagName":"header"} /-->`

> > 11. Use Create Block Theme plugin to ’Overwrite Parent’ 
> > 12. In a code editor observe that the index.html template in the parent theme is now changed, so it links to the template-part with
> > `<!-- wp:template-part {"slug":"header","theme":"parent","tagName":"header"} /-->`
> 13. Reactivate the child theme 
> 14. Observe that the child no longer loads the header template part from the parent theme
> 15. Open the index.html template in the site editor and observe the error message ’Template part has been deleted or is unavailable’.

Here the theme slug was automatically updated to that of the child theme and no error occurred:

`<!-- wp:template-part {"slug":"header","tagName":"header","theme":"child"} /-->`

> 16. In a code editor manually remove the theme key from the index.html template so the template-part link is restored to
> `<!-- wp:template-part {"slug":"header","tagName":"header"} /-->`
> 17. Observe that the child now correctly loads the header template part from the parent theme

Here I got the opposite result. If the theme slug was missing, a "Template part has been deleted or is unavailable: header" error occurred. However, this step should not be necessary since the expected results are obtained up to step 15.

**Comment by t-hamano** [comment]
I tested it again on WP6.5 Beta3, and the results are exactly the same as [this comment](https://github.com/WordPress/gutenberg/issues/44243#issuecomment-1826715470), and the problem appears to be completely resolved.

I'd like to close this issue, but if I've missed anything, please feel free to comment.

**Comment by realthemes** [comment]
My custom block theme keeps inserting ```"theme":"my-theme-name"``` in the navigation template part:

```<!-- wp:template-part {"slug":"rlt-navigation","theme":"my-theme-name","area":"header"} /-->```

This results in the error: "Template part has been deleted or is unavailable."

I'm unsure about the steps required to avoid this issue. I also suspect the Create Block Theme plugin might not be working correctly. It generates a child theme using API version 3, whereas my parent theme uses API version 2. Additionally, it doesn't create the single.html template, which I expect it to generate.

PS in theme.json file of the parent theme I have this code:
```
"templateParts": [
	{
		"area": "header",
		"name": "rlt-navigation",
		"title": "RLT Navigation Template Part"
	}
],
```

in parts folder of parent theme I have this file ```rlt-navigation.html``` with this code:
```
<!-- wp:rlt-blocks/rlt-navigation {"className":"is-style-rlt-navigation-shadow"} /-->
```
When I create child theme with Create Block Theme plugin, this results in the error: "Template part has been deleted or is unavailable." on this block

**Comment by t-hamano** [comment]
@realthemes Sorry for the late reply.

Does this problem only occur when you use the Create Block Theme plugin? If so, you may want to submit an issue in this repository: https://github.com/WordPress/create-block-theme

**Comment by realthemes** [comment]
@t-hamano, thank you for reply, I gave up on creating a child theme and using this plugin that time, so that's a thing of the past, but I'm creating a new block theme now, and when I get to the child theme stage, I'll try to reproduce this bug again and file a bug report.
